clippy_lints/src/async_yields_async.rs

```diff
@@ -87,13 +89,22 @@ impl<'tcx> LateLintPass<'tcx> for AsyncYieldsAsync {
         let expr_ty = typeck_results.expr_ty(body_expr);
 
         if implements_trait(cx, expr_ty, future_trait_def_id, &[]) {
-            let return_expr_span = match &body_expr.kind {
+            let return_expr = match &body_expr.kind {
                 // XXXkhuey there has to be a better way.
-                ExprKind::Block(block, _) => block.expr.map(|e| e.span),
-                ExprKind::Path(QPath::Resolved(_, path)) => Some(path.span),
+                ExprKind::Block(block, _) => block.expr.map(|e| (e, e.span)),
+                ExprKind::Path(QPath::Resolved(_, path)) => Some((body_expr, path.span)),
                 _ => None,
             };
-            if let Some(return_expr_span) = return_expr_span {
+            if let Some((return_expr, return_expr_span)) = return_expr {
```

Comment:
You may remove one level of indentation with an early return, and remove the need for `map()`:

```suggestion
            let (return_expr, return_expr_span) = match &body_expr.kind {
                ExprKind::Block(Block { expr: Some(e) ,..}, _) => (*e, e.span),
                ExprKind::Path(QPath::Resolved(_, path)) => (body_expr, path.span),
                _ => return,
            };
```